src/main/java/org/opensearch/neuralsearch/settings/NeuralSearchSettings.java

```diff
@@ -34,4 +33,49 @@ public final class NeuralSearchSettings {
         50,
         Setting.Property.NodeScope
     );
+
+    /**
+     * Use this setting to manage if a neuralSparseQuery build a two-phase query of not.
+     */
+    public static final Setting<Boolean> NEURAL_SPARSE_TWO_PHASE_DEFAULT_ENABLED = Setting.boolSetting(
```

Comment:
The reason for setting it as a node-scope setting is that we believe the feature's settings are relatively stable. In our initial investigations, we observed great performance and accuracy with the default settings. We also think that significant differences in two-phase parameters might cause variations in accuracy across different indexes. Therefore, we made this a node-scope setting. 
Additionally, for customization in a single query, we provide a one-time two-phase parameter that allows users to temporarily override this setting in an individual query."
```
{
  "query": {
    "neural_sparse":{
      "VECTOR_FIELD":
      {
        "model_id": "model_id",
        "query_text": "abcd",
        "two_phase_settings":
         {					
            "enabled": true
            "window_size_expansion": 5,  
            "pruning_ratio": 0.4,  
         }
      }
    }
  }
}
```

Reply:
if the settings are stable and we already provide user a way to set the settings in the query itself then we don't need to explicitly provide setting at cluster level because the way I look at this is these settings are of no use. They just taking space in cluster state. Hence I would recommend not having the settings. Plus more I look at this, these are not setting these are more of Neural Sparse query attributes. Hence, we should just remove the settings completely.

Reply:
The settings is stable for **a concrete sparse model**. If the sparse model is trained with [thresholding in loss function](https://dl.acm.org/doi/pdf/10.1145/3539618.3592051), they can have different hyperparameters to work with. While the default settings (5, 0.4) works best for our sparse models.

Let's make the decision for keeping these settings based on pros and cons:
pros:
- make it more convinient for users with customized sparse models
- make it more convinient to disable the 2-phase search, there are use cases to use OpenSearch to benchmark the sparse model. For these cases we need accurate search relevance scores
cons:
- having 3 more cluster settings, taking dozens of bytes in memory

Many settings in OpenSearch are rarely changed. Most users won't feel this settings, but for users who need the settings, it is very important to be able to configure them. I don't think this is a critical for performance, hence one single embedding vector will take much more space than these settings.

Reply:
If we are defining these settings based on the pros you mentioned then these settings are much more suited at index level and not at cluster level. Because if I have 2 indices using 2 different sparse models, then this setting will be of use. But if we make them at index level then atleast they are useful as I can change setting for 1 index. 

Moreover if you think little hard on that point, more than 1 model can be used in 2 different sparse fields on same index. In that case not even the index settings be useful. Hence if you see the settings are not very useful in that perspective because for most of the cases default works and for cases where default doesn't work user can update the query.

> cons:
having 3 more cluster settings, taking dozens of bytes in memory

Replicate this dozen of bytes over more than 100 nodes then you will see the impact. Nevertheless the point for that comment was, if the settings are not useful as I mentioned above then just adding them in cluster setting is just waste of resources. I cannot comment on other settings which customer don't change and still present in Opensearch. But all I will say is if something is not correct in current code is not a justification to follow the same pattern.

Reply:
We evaluated implementing them as index settings, and it will have performance issue.
For cluster settings we can use volatile object to just fetch the value from memory. But for index settings we must access the settings object and go through the [get setting API](https://github.com/opensearch-project/OpenSearch/blob/81707dcc74fd8a66349c80c0c47254de110b9f5d/server/src/main/java/org/opensearch/common/settings/Setting.java#L480), for every neural sparse search request

Reply:
If a setting is supposed to be an index setting and we making it as part of cluster settings just on the name of performance, do we know how much performance boost we are getting here ? Plus index settings are also stored on the node. So, when settings change you might get hit for 1 query but other queries will be fast.

Reply:
> If a setting is supposed to be an index setting and we making it as part of cluster settings just on the name of performance, do we know how much performance boost we are getting here?
The performance impact is very straightforward in code so we haven't benchmark it. For cluster settings, we just do value assignment; and for index settings, we need to iterate the environment settings. And this will happen for every neural sparse query.

> Plus index settings are also stored on the node. So, when settings change you might get hit for 1 query but other queries will be fast.
Do you mean we cache these settings for each index? We need to implement another complexed controller module to manage the cache for each index and do update even manage their lifecycle with index create/delete. I think this is completely overdesigned. You just mentioned the overhead to add one setting, implmenting index settings will make this overhead much larger because in every index we have these settings.

Reply:
> Do you mean we cache these settings for each index? We need to implement another complexed controller module to manage the cache for each index and do update even manage their lifecycle with index create/delete. I think this is completely overdesigned. You just mentioned the overhead to add one setting, implmenting index settings will make this overhead much larger because in every index we have these settings.

I am saying they are already cached at the node level. You don't need to cache it. So any time there is change in settings(index/cluster settings) for any index those settings are pushed to all the data nodes. 

> For cluster settings, we just do value assignment; and for index settings, we need to iterate the environment settings. And this will happen for every neural sparse query.

Not sure what you mean here.

Reply:
OK I get your point, I checked the settings API and found I misunderstand how it works. 
So you put +1 to index settings now?

Reply:
Changed to indexscope.